src/miral/wayland_extensions.cpp

```diff
@@ -339,15 +394,58 @@ void miral::WaylandExtensions::operator()(mir::Server& server) const
                 for (auto const& extension : dropped)
                 {
                     selected_extensions.erase(extension);
+                    manually_enabled_extensions.erase(extension);
+                    manually_disabled_extensions.insert(extension);
                 }
             }
 
+            for (auto const& pair : self->conditional_extensions)
+            {
+                selected_extensions.insert(pair.first);
+            }
+
             selected_extensions = map_to_global_names(selected_extensions);
+            manually_enabled_extensions = map_to_global_names(manually_enabled_extensions);
+            manually_disabled_extensions = map_to_global_names(manually_disabled_extensions);
+
             self->validate(selected_extensions);
             server.set_enabled_wayland_extensions(
                 std::vector<std::string>{
                     selected_extensions.begin(),
                     selected_extensions.end()});
+
+            server.set_wayland_extension_filter(
+                [&self = self, manually_enabled_extensions, manually_disabled_extensions]
+                (Application const& app, char const* protocol) -> bool
+                {
+                    // Wayland calls the filter for all protocols (not just the optional extensions). To avoid accidents
+                    // (like denying base protocols) we always accept extensions that are not in supported_extensions.
+                    if (self->supported_extensions.count(protocol) == 0)
+                    {
+                        return true;
+                    }
+                    auto const cond = self->conditional_extensions.find(protocol);
+                    if (cond != self->conditional_extensions.end())
+                    {
+                        std::optional<bool> user_pref;
+                        if (manually_enabled_extensions.find(protocol) != manually_enabled_extensions.end())
+                        {
+                            user_pref = true;
+                        }
+                        else if (manually_disabled_extensions.find(protocol) != manually_disabled_extensions.end())
+                        {
+                            user_pref = false;
+                        }
+                        return cond->second(EnableInfo{
+                            app,
+                            protocol,
+                            user_pref});
+                    }
+                    else
+                    {
+                        return self->extensions_filter ? self->extensions_filter.value()(app, protocol) : true;
+                    }
+                });
         });
```

Comment:
This lambda seems overcomplicated (as does the surrounding one). We can do a lot of the logic before calling `server.set_wayland_extension_filter()`:
```c++
            if (self->extensions_filter)
            {
                server.set_wayland_extension_filter(
                    [self](Application const& app, char const* protocol) -> bool
                    {
                        return self->extensions_filter.value()(app, protocol);
                    });
            }
            else if (!self->conditional_extensions.empty())
            {
                server.set_wayland_extension_filter(
                    [&self = self, manually_enabled_extensions, manually_disabled_extensions]
                    (Application const& app, char const* protocol) -> bool
                    {
                        auto const cond = self->conditional_extensions.find(protocol);

                        if (cond == self->conditional_extensions.end())
                        {
                            return true;
                        }
                        else
                        {
                            std::optional<bool> user_pref;
                            if (manually_enabled_extensions.find(protocol) != manually_enabled_extensions.end())
                            {
                                user_pref = true;
                            }
                            else if (manually_disabled_extensions.find(protocol) != manually_disabled_extensions.end())
                            {
                                user_pref = false;
                            }

                            return cond->second(EnableInfo{ app, protocol, user_pref});
                        }
                    });
            }
```

(Need to restore L369-373 too for that to work)